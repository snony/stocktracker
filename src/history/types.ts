export interface HistoryData {
  readonly [key: string]: string | number
  readonly date: string
  readonly price: number
}

export interface HistoryState {
  history: HistoryData[]
  dateFilter: string
  priceFilter: string
}

export interface HistoryChartProps {
  readonly history: HistoryData[]
}

export interface FilterButtonProps {
  readonly type: string
  readonly value: string
  readonly selected: boolean
  readonly onClick: (symbol: string, value: string) => void
}
