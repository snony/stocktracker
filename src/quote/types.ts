import { QUOTE } from 'socket/'
export interface QuoteProps {
    quote: QUOTE
    unsubscribe: () => void
}