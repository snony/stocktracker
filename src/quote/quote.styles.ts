import { css } from 'react-emotion'
import styled from 'styled'
export default styled('span')`
  display: flex;
  flex-direction: column nowrap;
  border-bottom: ${props => props.theme.searchBarHeading};
  padding: 0.7rem 0.5rem;
  width: 25rem;
  justify-content: space-evenly;
`

export const quoteClass = css`
  margin-bottom: 0;
  display: inline-block;
`
