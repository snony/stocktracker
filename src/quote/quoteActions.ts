import { Action, ActionCreator } from 'redux'
import { ThunkAction } from 'redux-thunk'
import { API, GlobalState } from 'types'

export type SymbolSubscriptionActions = SubscribeSymbolAction | UnsubscribeSymbolAction
interface SubscribeSymbolAction extends Action {
    type: typeof SYMBOL_SUBSCRIBE_ACTION
    symbol: string
}

interface UnsubscribeSymbolAction extends Action {
    type: typeof SYMBOL_UNSUBSCRIBE_ACTION
}
export const SYMBOL_SUBSCRIBE_ACTION = 'SYMBOL_SUBSCRIBE_ACTION'
export const subscribeSymbol: ActionCreator<SubscribeSymbolAction> = (symbol: string) => ({
    type: SYMBOL_SUBSCRIBE_ACTION,
    symbol
})

export const SYMBOL_UNSUBSCRIBE_ACTION = 'SYMBOL_UNSUBSCRIBE_ACTION'
export const unSubscribeSymbol: ActionCreator<UnsubscribeSymbolAction> = () => ({
    type: SYMBOL_UNSUBSCRIBE_ACTION
})

type ThunkResult = ThunkAction<void, GlobalState, API, SymbolSubscriptionActions>
export const subscribeSymbolThunk: (symbol: string) => ThunkResult = symbol => async (dispatch, _, ) => {
    return dispatch(subscribeSymbol(symbol))
}