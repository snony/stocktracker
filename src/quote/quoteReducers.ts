import { Reducer } from 'redux'

import {
  SYMBOL_SUBSCRIBE_ACTION,
  SYMBOL_UNSUBSCRIBE_ACTION,
  SymbolSubscriptionActions
} from './quoteActions'

export interface SymbolSubscriptionState {
  symbol: string
}
export const initState: SymbolSubscriptionState = { symbol: '' }

const subscribeSymbolReducers: Reducer<SymbolSubscriptionState, SymbolSubscriptionActions> = (
  state = initState,
  action
) => {
  switch (action.type) {
    case SYMBOL_SUBSCRIBE_ACTION:
      return { ...state, symbol: action.symbol }
    case SYMBOL_UNSUBSCRIBE_ACTION:
      return { ...state, symbol: '' }
    default:
      return state
  }
}

export default subscribeSymbolReducers
