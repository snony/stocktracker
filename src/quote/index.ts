import './quote.css'
import { setSubscribeSymbol, SYMBOL_SUBSCRIBE_ACTION, SYMBOL_UNSUBSCRIBE_ACTION } from './quoteActions'
import Quote from './quoteContainer'
import subscribeSymbolReducers from './quoteReducers'
export { setSubscribeSymbol, subscribeSymbolReducers, SYMBOL_UNSUBSCRIBE_ACTION, SYMBOL_SUBSCRIBE_ACTION }
export default Quote
