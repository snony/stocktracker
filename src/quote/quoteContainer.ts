import { connect } from 'react-redux'
import { ThunkDispatch } from 'redux-thunk'
import { QUOTE } from 'socket'
import { API, GlobalState } from 'types'

import Quote from './quote'
import { SymbolSubscriptionActions, unSubscribeSymbol } from './quoteActions'

export const processQuote = (quote: QUOTE, previousClose: number) => ({
    price: quote.lastSalePrice,
    change: quote.lastSalePrice - previousClose,
    changePercent: (Math.abs(quote.lastSalePrice - previousClose) / previousClose) * 100
})

export const mapStateToProps = (state: GlobalState) => ({
    quote: processQuote(state.quote, state.symbol.previousClose),
    fetchStatus: state.quote.fetchStatus
})

export type ThunkDispatchAction = ThunkDispatch<GlobalState, API, SymbolSubscriptionActions>
export const mapDispatchToProps = (dispatch: ThunkDispatchAction) => ({
    unsubscribe: () => dispatch(unSubscribeSymbol())
})

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Quote)
