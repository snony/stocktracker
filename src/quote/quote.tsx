import React from 'react'
import socketIo, { QUOTE } from '../socket'
import { QuoteProps } from './quoteContainer'

const initState = { lastSalePrice: 0, lastSaleSize: 0, marketPercent: 0 }

class Quote extends React.PureComponent<QuoteProps, QUOTE> {

    constructor(props: QuoteProps) {
        super(props)
        this.state = initState
    }

    public componentWillMount() {
        socketIo.connectSocket(this.props.company.symbol)
        socketIo.listenOnMessage(this.setQuote)
    }

    public componentDidUpdate(prevProps: QuoteProps) {
        if (prevProps.company.symbol !== this.props.company.symbol) {
            socketIo.emitUnsubscribe(prevProps.company.symbol)
            socketIo.emitSubscribeSymbol(this.props.company.symbol)
        }
    }

    public componentWillUnmount() {
        socketIo.emitUnsubscribe(this.props.company.symbol)
        socketIo.disconnectSocket()
    }

    public render() {
        return (
            <div className='search_container___quote_wrapper'>
                <span className='label'>{this.state.lastSalePrice} </span>
                <span className='label'> {this.state.lastSaleSize} </span>
                <span className='label'> {this.state.marketPercent} </span>
            </div>
        )
    }

    private setQuote = (data: QUOTE) => {
        this.setState({ ...data })
    }
}


export default Quote