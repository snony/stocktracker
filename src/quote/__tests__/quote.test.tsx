import Quote from 'quote/quote'
import { QuoteProps } from 'quote/types';
import React from 'react'
import renderer from 'react-test-renderer'

const getQuote = (partialQuote: Partial<QuoteProps>) => ({
    quote: { lastSalePrice: 0, lastSaleSize: 0, marketPercent: 0 },
    unsubscribe: jest.fn(),
    ...partialQuote
})

describe('Quote Component', () => {

    it('renders correctly on initial state', () => {
        const quoteData = getQuote({})
        const tree = renderer.create(<Quote {...quoteData} />).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('renders correctly with updated values', () => {
        const quoteData = getQuote({ quote: { lastSalePrice: 20, lastSaleSize: 1, marketPercent: 0.5 } })
        const tree = renderer.create(<Quote {...quoteData} />).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('should call unsubscribe function on unmount', () => {
        const unsubscribe = jest.fn()
        const quoteData = getQuote({ unsubscribe })
        const tree = renderer.create(<Quote {...quoteData} />)
        tree.unmount()
        expect(unsubscribe).toHaveBeenCalledTimes(1)

    })
})