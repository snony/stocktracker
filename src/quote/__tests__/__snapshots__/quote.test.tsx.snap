// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote Component renders correctly on initial state 1`] = `
<div
  className="search_container___quote_wrapper"
>
  <span
    className="label"
  >
    0
     
  </span>
  <span
    className="label"
  >
     
    0
     
  </span>
  <span
    className="label"
  >
     
    0
     
  </span>
</div>
`;

exports[`Quote Component renders correctly with updated values 1`] = `
<div
  className="search_container___quote_wrapper"
>
  <span
    className="label"
  >
    20
     
  </span>
  <span
    className="label"
  >
     
    1
     
  </span>
  <span
    className="label"
  >
     
    0.5
     
  </span>
</div>
`;
